echo_server.c

```diff
@@ -49,65 +57,109 @@ static int send_request(struct socket *sock, unsigned char *buf, size_t size)
 
     printk(MODULE_NAME ": start send request.\n");
 
+    /* FIXME: what if somehow length of message is zero? */
     length = kernel_sendmsg(sock, &msg, &vec, 1, strlen(buf) - 1);
 
     printk(MODULE_NAME ": send request = %s\n", buf);
 
     return length;
 }
 
-static int echo_server_worker(void *arg)
+static void echo_server_worker(struct work_struct *work)
 {
-    struct socket *sock;
+    struct fastecho *worker =
+        container_of(work, struct fastecho, fastecho_work);
     unsigned char *buf;
     int res;
 
-    sock = (struct socket *) arg;
-    allow_signal(SIGKILL);
-    allow_signal(SIGTERM);
-
     buf = kmalloc(BUF_SIZE, GFP_KERNEL);
     if (!buf) {
         printk(KERN_ERR MODULE_NAME ": kmalloc error....\n");
-        return -1;
+        return;
     }
 
-    while (!kthread_should_stop()) {
-        res = get_request(sock, buf, BUF_SIZE - 1);
+    /*
+     * FIXME: first echo message may contains unexpected characters
+     * (you may even get kernel cmdline)
+     */
+
+    /*
+     * currently, module removal task will be blocked before
+     * closure of existing connection, better idea?
```

Comment:
You shall take pending requests into consideration.

Reply:
Do you mean `get_request()` keep waiting message from client?

Reply:
sort of. Client requests need to be treated explicitly.

Reply:
The module can now being removed (`rmmod`) smoothly. We shutdown all sockets and release them at cleanup of the daemon. And besides, we added another `kernel_setsockopt()` inside `open_listen()`, which sets `SO_REUSEPORT` option to prevent `Address already in use` error during binding the socket.

In `fastecho_module.c`:
```c
error = kernel_setsockopt(sock, SOL_SOCKET, SO_REUSEPORT, (char *) &opt,
                              sizeof(opt));                    
if (error < 0) {
    printk(KERN_ERR MODULE_NAME
              ": setsockopt SO_REUSEPORT setting error = %d\n",
               error);
    sock_release(sock);
    return error;
}
```

In `echo_server.c`:
```c
static void free_work(void)
{
    struct fastecho *l = NULL, *tar = NULL;

    list_for_each_entry_safe (tar, l, &daemon.worker, list) {
        kernel_sock_shutdown(tar->sock, SHUT_RDWR);
        sock_release(tar->sock);
        flush_work(&tar->fastecho_work);
        kfree(tar);
    }
}
```

We were trying to figure out why a connection force closed (same operations as normal closure performed) by the daemon would be left in `ESTABLISHED` state. Turns out if a connection didn't close properly (4-way-handshake), we can close it by sending a kill signal, but here, the process is kernel process, we can't send it a kill. Hence, we think the solution above maybe is the most we can do.

Connection info printed by `netstat`:
```
tcp        0      0 127.0.0.1:12345         127.0.0.1:50324         ESTABLISHED -
```